sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala

```diff
@@ -88,6 +88,16 @@ abstract class BinaryArithmetic extends BinaryOperator {
 
   override def dataType: DataType = left.dataType
 
+  // Whether the dataType is promoted or not, we should only promote DecimalType once
+  var decimalPromoted: Boolean = false
```

Comment:
Instead of using a cast can we create an explicit expression for changing decimal precision and then the match will only apply if this operator is not already present?